--- a/kernel/feature/kernel_umount.c
+++ b/kernel/feature/kernel_umount.c
@@ -83,6 +83,7 @@
 		return 0;
 	}
 
+#ifndef CONFIG_KSU_SUSFS
 	// There are 6 scenarios:
 	// 1. Normal app: zygote -> appuid
 	// 2. Isolated process forked from zygote: zygote -> isolated_process
@@ -107,6 +108,8 @@
 		pr_info("handle umount ignore non zygote child: %d\n", current->pid);
 		return 0;
 	}
+#endif
+
 	// umount the target mnt
 	pr_info("handle umount for uid: %d, pid: %d\n", new_uid, current->pid);
 
@@ -125,14 +128,16 @@
 	return 0;
 }
 
-void __init ksu_kernel_umount_init(void)
+// kernel_umount: when a process is setuid, if it is an app, we umount all modules
+// for it.
+void __init ksu_kernel_umount_init()
 {
 	if (ksu_register_feature_handler(&kernel_umount_handler)) {
 		pr_err("Failed to register kernel_umount feature handler\n");
 	}
 }
 
-void __exit ksu_kernel_umount_exit(void)
+void __exit ksu_kernel_umount_exit()
 {
 	ksu_unregister_feature_handler(KSU_FEATURE_KERNEL_UMOUNT);
 }
